Cluster stats endpoint broken with ES 7

## Description

The following API endpoints are currently broken with ES 7:

- `/api/system/cluster/stats`
- `/api/system/cluster/stats/elasticsearch`

<details>
<summary>Log messages</summary>

```
2020-11-10 23:19:17,281 WARN : org.graylog.shaded.elasticsearch7.org.elasticsearch.client.RestClient - request [GET http://127.0.0.1:9200/_cluster/stats/nodes] returned 1 warnings: [299 Elasticsearch-7.9.3-c4138e51121ef06a6404866cddc601906fe5c868 "[types removal] Specifying types in document get requests is deprecated, use the /{index}/_doc/{id} endpoint instead."]
2020-11-10 23:19:17,282 ERROR: org.graylog2.shared.rest.exceptionmappers.AnyExceptionClassMapper - Unhandled exception in REST resource
org.graylog.shaded.elasticsearch7.org.elasticsearch.ElasticsearchException: Couldn't read Elasticsearch cluster stats
	at org.graylog.storage.elasticsearch7.ElasticsearchClient.exceptionFrom(ElasticsearchClient.java:106) ~[classes/:?]
	at org.graylog.storage.elasticsearch7.ElasticsearchClient.execute(ElasticsearchClient.java:76) ~[classes/:?]
	at org.graylog.storage.elasticsearch7.PlainJsonApi.perform(PlainJsonApi.java:22) ~[classes/:?]
	at org.graylog.storage.elasticsearch7.ClusterAdapterES7.clusterStats(ClusterAdapterES7.java:197) ~[classes/:?]
	at org.graylog2.indexer.cluster.Cluster.elasticsearchStats(Cluster.java:192) ~[classes/:?]
	at org.graylog2.system.stats.ClusterStatsService.elasticsearchStats(ClusterStatsService.java:101) ~[classes/:?]
	at org.graylog2.system.stats.ClusterStatsService.clusterStats(ClusterStatsService.java:78) ~[classes/:?]
	at org.graylog2.rest.resources.system.ClusterStatsResource.systemStats(ClusterStatsResource.java:52) ~[classes/:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_272]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_272]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_272]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_272]
	at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory.lambda$static$0(ResourceMethodInvocationHandlerFactory.java:52) ~[jersey-server-2.31.jar:?]
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:124) ~[jersey-server-2.31.jar:?]
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:167) ~[jersey-server-2.31.jar:?]
	at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$TypeOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:219) ~[jersey-server-2.31.jar:?]
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:79) ~[jersey-server-2.31.jar:?]
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:469) ~[jersey-server-2.31.jar:?]
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:391) ~[jersey-server-2.31.jar:?]
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:80) ~[jersey-server-2.31.jar:?]
	at org.glassfish.jersey.server.ServerRuntime$1.run(ServerRuntime.java:253) [jersey-server-2.31.jar:?]
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:248) [jersey-common-2.31.jar:?]
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:244) [jersey-common-2.31.jar:?]
	at org.glassfish.jersey.internal.Errors.process(Errors.java:292) [jersey-common-2.31.jar:?]
	at org.glassfish.jersey.internal.Errors.process(Errors.java:274) [jersey-common-2.31.jar:?]
	at org.glassfish.jersey.internal.Errors.process(Errors.java:244) [jersey-common-2.31.jar:?]
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:265) [jersey-common-2.31.jar:?]
	at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:232) [jersey-server-2.31.jar:?]
	at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:680) [jersey-server-2.31.jar:?]
	at org.glassfish.jersey.grizzly2.httpserver.GrizzlyHttpContainer.service(GrizzlyHttpContainer.java:356) [jersey-container-grizzly2-http-2.31.jar:?]
	at org.glassfish.grizzly.http.server.HttpHandler$1.run(HttpHandler.java:200) [grizzly-http-server-2.4.4.jar:2.4.4]
	at com.codahale.metrics.InstrumentedExecutorService$InstrumentedRunnable.run(InstrumentedExecutorService.java:180) [metrics-core-4.1.9.jar:4.1.9]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_272]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_272]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_272]
Caused by: org.graylog.shaded.elasticsearch7.org.elasticsearch.client.ResponseException: method [GET], host [http://127.0.0.1:9200], URI [/_cluster/stats/nodes], status line [HTTP/1.1 404 Not Found]
Warnings: [[types removal] Specifying types in document get requests is deprecated, use the /{index}/_doc/{id} endpoint instead.]
{"error":{"root_cause":[{"type":"index_not_found_exception","reason":"no such index [_cluster]","resource.type":"index_expression","resource.id":"_cluster","index_uuid":"_na_","index":"_cluster"}],"type":"index_not_found_exception","reason":"no such index [_cluster]","resource.type":"index_expression","resource.id":"_cluster","index_uuid":"_na_","index":"_cluster"},"status":404}
	at org.graylog.shaded.elasticsearch7.org.elasticsearch.client.RestClient.convertResponse(RestClient.java:302) ~[elasticsearch7-7.9.1-0.jar:?]
	at org.graylog.shaded.elasticsearch7.org.elasticsearch.client.RestClient.performRequest(RestClient.java:272) ~[elasticsearch7-7.9.1-0.jar:?]
	at org.graylog.shaded.elasticsearch7.org.elasticsearch.client.RestClient.performRequest(RestClient.java:246) ~[elasticsearch7-7.9.1-0.jar:?]
	at org.graylog.storage.elasticsearch7.PlainJsonApi.lambda$perform$0(PlainJsonApi.java:24) ~[classes/:?]
	at org.graylog.storage.elasticsearch7.ElasticsearchClient.execute(ElasticsearchClient.java:74) ~[classes/:?]
	... 33 more
```

</details>

--- a/UPGRADING.rst
+++ b/UPGRADING.rst
@@ -4,6 +4,15 @@ Upgrading to Graylog 4.0.x
 
 .. _upgrade-from-33-to-40:
 
+Deprecation of cluster stats endpoints
+======================================
+
+Starting with v4.0, the cluster stats endpoints are deprecated and will be removed in a future version. Those include:
+
+  - '/system/cluster/stats'
+  - '/system/cluster/stats/elasticsearch'
+  - '/system/cluster/stats/mongo'
+
 [BREAKING] Fixing certificate validation for LDAP servers used for authentication
 =================================================================================
 

--- a/graylog-storage-elasticsearch7/src/main/java/org/graylog/storage/elasticsearch7/ClusterAdapterES7.java
+++ b/graylog-storage-elasticsearch7/src/main/java/org/graylog/storage/elasticsearch7/ClusterAdapterES7.java
@@ -192,7 +192,7 @@ public class ClusterAdapterES7 implements ClusterAdapter {
 
     @Override
     public ClusterStats clusterStats() {
-        final Request request = new Request("GET", "/_cluster/stats/nodes");
+        final Request request = new Request("GET", "/_cluster/stats");
 
         final JsonNode clusterStatsResponseJson = jsonApi.perform(request,
                 "Couldn't read Elasticsearch cluster stats");

--- a/graylog-storage-elasticsearch7/src/main/java/org/graylog/storage/elasticsearch7/ClusterAdapterES7.java
+++ b/graylog-storage-elasticsearch7/src/main/java/org/graylog/storage/elasticsearch7/ClusterAdapterES7.java
@@ -192,7 +192,7 @@ public class ClusterAdapterES7 implements ClusterAdapter {
 
     @Override
     public ClusterStats clusterStats() {
-        final Request request = new Request("GET", "/_cluster/stats");
+        final Request request = new Request("GET", "/_cluster/stats/nodes/*");
 
         final JsonNode clusterStatsResponseJson = jsonApi.perform(request,
                 "Couldn't read Elasticsearch cluster stats");

--- a/graylog2-server/src/main/java/org/graylog/plugins/views/search/views/DashboardService.java
+++ b/graylog2-server/src/main/java/org/graylog/plugins/views/search/views/DashboardService.java
@@ -30,7 +30,7 @@ public class DashboardService {
     }
 
     public long count() {
-        final PaginatedList<ViewDTO> result = viewService.searchPaginatedByType(ViewDTO.Type.DASHBOARD, new SearchQuery(""), dashboard -> true, "ASC", null, 1, 0);
+        final PaginatedList<ViewDTO> result = viewService.searchPaginatedByType(ViewDTO.Type.DASHBOARD, new SearchQuery(""), dashboard -> true, "ASC", ViewDTO.FIELD_ID, 1, 0);
         return result.grandTotal().orElseThrow(() -> new IllegalStateException("Missing grand total in response when counting dashboards!"));
     }
 }

--- a/graylog2-server/src/main/java/org/graylog/plugins/views/search/views/ViewService.java
+++ b/graylog2-server/src/main/java/org/graylog/plugins/views/search/views/ViewService.java
@@ -39,6 +39,7 @@ import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
 import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
 import static com.google.common.base.Strings.isNullOrEmpty;
 
 public class ViewService extends PaginatedDbService<ViewDTO> {
@@ -90,6 +91,7 @@ public class ViewService extends PaginatedDbService<ViewDTO> {
                                                         String sortField,
                                                         int page,
                                                         int perPage) {
+        checkNotNull(sortField);
         return searchPaginated(
                 DBQuery.and(
                         DBQuery.or(DBQuery.is(ViewDTO.FIELD_TYPE, type), DBQuery.notExists(ViewDTO.FIELD_TYPE)),

--- a/graylog2-server/src/main/java/org/graylog2/rest/resources/system/ClusterStatsResource.java
+++ b/graylog2-server/src/main/java/org/graylog2/rest/resources/system/ClusterStatsResource.java
@@ -32,10 +32,11 @@ import javax.ws.rs.Path;
 import javax.ws.rs.Produces;
 import javax.ws.rs.core.MediaType;
 
-@Api(value = "System/ClusterStats", description = "Cluster stats")
+@Api(value = "System/ClusterStats", description = "[DEPRECATED] Cluster stats")
 @RequiresAuthentication
 @Path("/system/cluster/stats")
 @Produces(MediaType.APPLICATION_JSON)
+@Deprecated
 public class ClusterStatsResource extends RestResource {
     private final ClusterStatsService clusterStatsService;
 

--- a/graylog2-server/src/main/java/org/graylog2/system/stats/ClusterStatsService.java
+++ b/graylog2-server/src/main/java/org/graylog2/system/stats/ClusterStatsService.java
@@ -28,7 +28,6 @@ import org.graylog2.streams.StreamService;
 import org.graylog2.system.stats.elasticsearch.ElasticsearchStats;
 import org.graylog2.system.stats.mongo.MongoProbe;
 import org.graylog2.system.stats.mongo.MongoStats;
-import org.graylog2.users.RoleService;
 
 import javax.inject.Inject;
 import javax.inject.Singleton;
@@ -42,7 +41,6 @@ public class ClusterStatsService {
     private final StreamService streamService;
     private final StreamRuleService streamRuleService;
     private final OutputService outputService;
-    private final RoleService roleService;
     private final AlertService alertService;
     private final AlarmCallbackConfigurationService alarmCallbackConfigurationService;
     private final DashboardService dashboardService;
@@ -55,7 +53,6 @@ public class ClusterStatsService {
                                StreamService streamService,
                                StreamRuleService streamRuleService,
                                OutputService outputService,
-                               RoleService roleService,
                                AlertService alertService,
                                AlarmCallbackConfigurationService alarmCallbackConfigurationService,
                                DashboardService dashboardService,
@@ -66,7 +63,6 @@ public class ClusterStatsService {
         this.streamService = streamService;
         this.streamRuleService = streamRuleService;
         this.outputService = outputService;
-        this.roleService = roleService;
         this.alertService = alertService;
         this.alarmCallbackConfigurationService = alarmCallbackConfigurationService;
         this.dashboardService = dashboardService;

